[transaction-replay] Implement a few commands for txn-replay tool

runtian-zhou opened the pull request:
…ove CLI

<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

Implement three commands:
1.  An option to save account states to local storage so that they can be later verified by the Move CLI.
2. Refactored the logic for executing unsigned writeset for verification purpose, and integrate such functionality to txn-replay tool
3. A command to fetch all the key accounts on chain: including core code address, diem_root, treasuary_compliance and all validators.
4. A checked-in release artifact along with a `verify-release` for `writeset-generator` to make sure the blob generated is indeed built from the checked in artifact.

I also updated these commands in the release process runbook.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

1. Tested the `annotate-account` command locally for both core_code_address and assoc_root address.
2. Tested the `writeset-generator`'s release command still works, and the generated blob can be re-executed by the txn replay tool.
3. Tested the `annotate-key-accounts` with premainnet and make sure states are fetched and passes the `move-doctor`
4. Tested the generated blob from `create-release` command for `writeset-generator` can pass the `verify-release` command. The command for creating a release and verifying the release will now look like following:
`cargo run -- --output out.bin --output-payload create-release PREMAINNET <BLOCK_HEIGHT>`
`cargo run -- verify-release PREMAINNET out.bin` 

## Related PRs

#7165

mengxu-fb commented on .gitignore:
```diff
@@ -51,3 +52,4 @@ target-out-docker
 
 # Doc generation output
 *.md.old
+/language/diem-tools/transaction-replay/storage/
```
Same as line 41?

runtian-zhou replied:
oops. My bad

runtian-zhou commented:
Update: Fixed Sam's comment in the first commit. Add a second commit to implement a new command for `writeset-generator` to checkin and verify the release artifact.

sblackshear commented on language/diem-tools/diem-validator-interface/src/lib.rs:
```diff
@@ -53,6 +50,48 @@ pub trait DiemValidatorInterface {
         Ok(acc)
     }
+
+    fn get_key_accounts_by_version(
```
Nit: `get_admin_accounts`, `get_governance_accounts` or similar might be better? The term "key" is so overloaded that I'd prefer not to use it...

sblackshear replied:
Also, a comment saying what this does seems useful.

runtian-zhou replied:
Fixed!

sblackshear commented on language/diem-tools/writeset-transaction-generator/src/main.rs:
```diff
@@ -47,6 +51,50 @@ enum Command {
+        writeset_path: PathBuf,
+        /// True if the serialized byte has the type of WriteSetPayload.
+        #[structopt(long)]
+        is_payload: bool,
```
This seems funky. It would probably be simpler not to allow polymorphism in the type of the blob. Is there a reason this tries to support validating both a WriteSet transaction and a payload?

runtian-zhou replied:
The main problem here is this: writeset used for AOS has different type from the writeset used for waypoint. That's why we need the output-payload flag and the is-payload flag here. Although this command is mostly used for the AOS for release process, in rare cases where we want to roll out emergency changes via waypoint this flag will be needed. E.g: if we find a critical secruity breach in Move modules on chain

runtian-zhou replied:
I replaced this flag with a either or logic with reasons given above: we want to be able to verify both types because in extreme cases we might want to create a release when the system is down, then we have to rely on the Waypoint Transaction to bring us back.

runtian-zhou replied:
Another question is should we distinguish the two types of payload by enforcing a extension name for a blob? That we would be able to get rid of the polymorphism over there.

sblackshear replied:
I like the extension-based approach.

sblackshear commented on language/diem-tools/writeset-transaction-generator/src/main.rs:
```diff
@@ -47,6 +51,50 @@ enum Command {
+/// Checkin the release artifact to a local file to make sure the release process is reproducible
+#[derive(Serialize, Deserialize)]
+struct ReleaseArtifact {
+    remote_json_rpc_url: String,
```
I think it would be more future-proof to use a ChainID instead of URL (both here and in the inputs to the CLI). The problem: the URLs are not stable identifiers (e.g., they have changed several times already due to the Diem renaming), whereas a chain ID is a stable identifier by definition.

sblackshear commented on language/diem-tools/writeset-transaction-generator/src/main.rs:
```diff
@@ -69,24 +117,64 @@ fn main() -> Result<()> {
-            let payload = create_release_writeset(remote_modules, local_modules)?;
-            verify_payload_change(&remote, Some(version), &payload)?;
+            if let Ok(artifact) = load_artifact() {
+                if artifact.remote_json_rpc_url != url || artifact.version > version {
```
I think this will fail if we try to create releases for multiple deployments from the same git branch. We probably want `load_artifact` to take a `chain_id` or `url` + have separate directories for release artifacts of each deployment.

runtian-zhou replied:
That's right. Lemme change that

runtian-zhou commented:
Upadte: Replace the json_rpc url with a ChainID with a builtin url mapping and save the artifact file per ChainID. Also got rid of the `is_payload` flag by using a or_else logic.

sblackshear reviewed (approved):
LGTM. 

Not needed for this PR, but in the future we should create some `ValidatorInterface` mocks that will allow us to test all these flows...

runtian-zhou commented:
> LGTM.
> 
> Not needed for this PR, but in the future we should create some `ValidatorInterface` mocks that will allow us to test all these flows...

I've been thinking about this for a while. To test these flows we really need a mock node setup (like `diem-swarm`) locally so that we have a working json-rpc endpoint we can talk to. However we need to create some changes to the Diem Framework to test this flow and that part is non-trivial for me to automate.

sblackshear commented:
One idea for that: you can have a for-testing-only copy of a very stable file (e.g., `Signature.move`) that contains a small change or two. The test can copy the changed file in, make a few other easy changes (e.g., delete a module, add a new module) rebuild everything, and make sure it works.

sblackshear commented on language/diem-tools/writeset-transaction-generator/src/main.rs:
```diff
@@ -68,25 +126,78 @@ fn main() -> Result<()> {
+            url,
+            version,
+        } => {
+            if let Ok(artifact) = load_artifact(&chain_id) {
```
It might be better to use `load_artifact?` here and manually create release `version` 0 artifacts for each chain ID?

runtian-zhou replied:
Added a maunal flag to create the artifact for the first release.

sblackshear commented on language/diem-tools/writeset-transaction-generator/src/main.rs:
```diff
@@ -42,11 +47,64 @@ enum Command {
+
+/// Checkin the release artifact to a local file to make sure the release process is reproducible
+#[derive(Serialize, Deserialize)]
+struct ReleaseArtifact {
```
I'm wondering if we also want something that commits to the framework code (e.g., a hash) here. Thoughts?

runtian-zhou replied:
A hash of the current stdlib would be nice. Will add that.

runtian-zhou replied:
PR Updated in fixup commit!

sblackshear commented on language/diem-tools/writeset-transaction-generator/src/main.rs:
```diff
@@ -42,11 +47,64 @@ enum Command {
+#[derive(Serialize, Deserialize)]
+struct ReleaseArtifact {
+    chain_id: ChainId,
+    // TODO: Deprecate after release
```
Do we need to keep this?

runtian-zhou replied:
I thought keeping this item will be desirable because we can prevent the manual error from inputting a chain_id with a different url.

runtian-zhou commented:
/land

runtian-zhou commented:
/abort

bors-libra commented:
:exclamation: Invalid command

<details>
<summary>Bors help and documentation</summary>
<br />

Bors is a Github App used to manage merging of PRs in order to ensure that CI is always green. It does so by maintaining a [Merge Queue](https://github.com/diem/diem/projects/2). Once a PR reaches the head of the Merge Queue it is rebased on top of the latest version of the PR's `base-branch` (generally `master`) and then triggers CI. If CI comes back green the PR is then merged into the `base-branch`. Regardless of the outcome, the next PR is the queue is then processed.

### General
- This project's Merge Queue can be found [here](https://github.com/diem/diem/projects/2).
- This project requires PRs to be [Reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews) and Approved before they can be queued for merging.
- Before PR's can be merged they must be configured with ["Allow edits from maintainers"](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) enabled. This is needed for Bors to be able to update PR's in-place so that Github can properly recognize and mark them as "merged" once they've been merged into the upstream branch.
### Commands
Bors actions can be triggered by posting a comment which includes a line of the form `/<action>`.
| Command | Action | Description |
| --- | --- | --- |
| __Land__ | `land`, `merge` | attempt to land or merge a PR |
| __Canary__ | `canary`, `try` | canary a PR by performing all checks without merging |
| __Cancel__ | `cancel`, `stop` | stop an in-progress land |
| __Cherry Pick__ | `cherry-pick <target>` | cherry-pick a PR into `<target>` branch |
| __Priority__ | `priority` | set the priority level for a PR (`high`, `normal`, `low`) |
| __Help__ | `help`, `h` | show this help message |

### Options
Options for Pull Requests are configured through the application of labels.
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Option&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Description |
| --- | --- |
| ![label: bors-high-priority](https://img.shields.io/static/v1?label=&message=bors-high-priority&color=lightgrey) | Indicates that the PR is high-priority. When queued the PR will be placed at the head of the merge queue. |
| ![label: bors-low-priority](https://img.shields.io/static/v1?label=&message=bors-low-priority&color=lightgrey) | Indicates that the PR is low-priority. When queued the PR will be placed at the back of the merge queue. |
| ![label: bors-squash](https://img.shields.io/static/v1?label=&message=bors-squash&color=lightgrey) | Before merging the PR will be squashed down to a single commit, only retaining the commit message of the first commit in the PR. |

</details>

runtian-zhou commented:
/cancel

runtian-zhou commented:
I'm working on a e2e test of this flow in smoke test. Will put it out in a separate PR.

runtian-zhou commented:
Updated per @sblackshear requests:
1. Added hash for the aritfact
2. Got rid of the URL in artifact
3. Change aritfact to json for better readability
4. Create a `first-release` option to create the first release artifact on disk. If the flag is not set the release process will abort if no artifact file is found.

runtian-zhou commented:
/land

runtian-zhou commented:
/cancel

bors-libra commented:
:exclamation: Land has been canceled due to this PR being updated with new commits. Please issue another Land command if you want to requeue this PR.

runtian-zhou commented:
/land
